src/bucket/BucketList.cpp

```diff
@@ -550,20 +573,67 @@ BucketList::addBatch(Application& app, uint32_t currLedger,
 }
 
 void
-BucketList::restartMerges(Application& app, uint32_t maxProtocolVersion)
+BucketList::restartMerges(Application& app, uint32_t maxProtocolVersion,
+                          uint32_t ledgerSeq, bool useFutureBucketsFromHAS)
 {
+    // Only possible with the new protocol
+    if (maxProtocolVersion < Bucket::FIRST_PROTOCOL_SHADOWS_REMOVED &&
+        !useFutureBucketsFromHAS)
+    {
+        throw std::runtime_error(
+            "Must re-start old-style merges (pre-protocol 12)");
+    }
+
+    std::shared_ptr<Bucket> snap;
     for (uint32_t i = 0; i < static_cast<uint32>(mLevels.size()); i++)
     {
         auto& level = mLevels[i];
         auto& next = level.getNext();
-        if (next.hasHashes() && !next.isLive())
+
+        if (useFutureBucketsFromHAS)
+        {
+            if (next.hasHashes() && !next.isLive())
+            {
+                next.makeLive(app, maxProtocolVersion, i);
+                if (next.isMerging())
+                {
+                    CLOG(INFO, "Bucket")
+                        << "Restarted merge on BucketList level " << i;
+                }
+            }
+        }
+        else
         {
-            next.makeLive(app, maxProtocolVersion, i);
-            if (next.isMerging())
+            // As per C++14 standard, both shared_future (async return object)
+            // and packaged_task (async provider) hold references to the shared
+            // state. On `clear`, shared_future releases its reference to the
+            // shared state without blocking. Upon finishing, packaged_task (for
+            // now irrelevant merge) will mark shared state as ready, and will
+            // then release its reference to shared state.
+
+            next.clear();
+            if (i > 0)
             {
-                CLOG(INFO, "Bucket")
-                    << "Restarted merge on BucketList level " << i;
+                // Before the last spill
```

Comment:
I know we've been over this and you convinced me before, but I'm forgetful and nervous and would love to see spelled-out in a comment here why merging an empty snap and a sometimes-empty curr is always right here. Or in general just "why this code is a correct restart of a merge".